pilot/pkg/model/config_test.go

```diff
@@ -282,6 +288,63 @@ func TestMostSpecificHostMatch(t *testing.T) {
 	}
 }
 
+func BenchmarkMostSpecificHostMatchExact(b *testing.B) {
+	length := 5000
```

Comment:
You should not only test the case of 5000 entries, also need to consider the case of 99% of users were there are only a few.

Reply:
In particular, simple array lookup is generally far faster than map construction+lookup for small input sizes

Reply:
ok, I will write more testcase use table driven to consider more cases. thanks for RC!

Reply:
Benchmark updated. I used the following benchmark in the original branch：
```go
func BenchmarkMostSpecificHostMatch(b *testing.B) {
	benchmarks := []struct {
		name     string
		needle   host.Name
		baseHost string
		hosts    []host.Name
		time     int
	}{
		{"10Exact", host.Name("foo.bar.com.10"), "foo.bar.com", []host.Name{}, 10},
		{"50Exact", host.Name("foo.bar.com.50"), "foo.bar.com", []host.Name{}, 50},
		{"100Exact", host.Name("foo.bar.com.100"), "foo.bar.com", []host.Name{}, 100},
		{"1000Exact", host.Name("foo.bar.com.1000"), "foo.bar.com", []host.Name{}, 1000},
		{"5000Exact", host.Name("foo.bar.com.5000"), "foo.bar.com", []host.Name{}, 5000},

		{"10DestRuleWildcard", host.Name("foo.bar.com.10"), "*.foo.bar.com", []host.Name{}, 10},
		{"50DestRuleWildcard", host.Name("foo.bar.com.50"), "*.foo.bar.com", []host.Name{}, 50},
		{"100DestRuleWildcard", host.Name("foo.bar.com.100"), "*.foo.bar.com", []host.Name{}, 100},
		{"1000DestRuleWildcard", host.Name("foo.bar.com.1000"), "*.foo.bar.com", []host.Name{}, 1000},
		{"5000DestRuleWildcard", host.Name("foo.bar.com.5000"), "*.foo.bar.com", []host.Name{}, 5000},

		{"10NeedleWildcard", host.Name("*.bar.foo.bar.com"), "*.foo.bar.com", []host.Name{}, 10},
		{"50NeedleWildcard", host.Name("*.bar.foo.bar.com"), "*.foo.bar.com", []host.Name{}, 50},
		{"100NeedleWildcard", host.Name("*.bar.foo.bar.com"), "*.foo.bar.com", []host.Name{}, 100},
		{"1000NeedleWildcard", host.Name("*.bar.foo.bar.com"), "*.foo.bar.com", []host.Name{}, 1000},
		{"5000NeedleWildcard", host.Name("*.bar.foo.bar.com"), "*.foo.bar.com", []host.Name{}, 5000},
	}

	for _, bm := range benchmarks {
		for i := 1; i <= bm.time; i++ {
			h := host.Name(bm.baseHost + "." + strconv.Itoa(i))
			bm.hosts = append(bm.hosts, h)
		}

		b.Run(bm.name, func(b *testing.B) {
			for n := 0; n < b.N; n++ {
				_, _ = model.MostSpecificHostMatch(bm.needle, bm.hosts)
			}
		})
	}

}
```
Here are some differences.

Before use map:
```shell
➜ go test ./pilot/pkg/model/config_test.go -test.run MostSpecificHostMatch -bench=.
goos: darwin
goarch: amd64
BenchmarkMostSpecificHostMatch/10Exact-8         	30014857	        39.5 ns/op
BenchmarkMostSpecificHostMatch/50Exact-8         	 2855409	       420 ns/op
BenchmarkMostSpecificHostMatch/100Exact-8        	 3138928	       383 ns/op
BenchmarkMostSpecificHostMatch/1000Exact-8       	  313680	      3767 ns/op
BenchmarkMostSpecificHostMatch/5000Exact-8       	   29348	     45219 ns/op
BenchmarkMostSpecificHostMatch/10DestRuleWildcard-8         	15788020	        75.9 ns/op
BenchmarkMostSpecificHostMatch/50DestRuleWildcard-8         	 4499768	       266 ns/op
BenchmarkMostSpecificHostMatch/100DestRuleWildcard-8        	 1697108	       707 ns/op
BenchmarkMostSpecificHostMatch/1000DestRuleWildcard-8       	  172573	      6955 ns/op
BenchmarkMostSpecificHostMatch/5000DestRuleWildcard-8       	   46590	     25882 ns/op
BenchmarkMostSpecificHostMatch/10NeedleWildcard-8           	14718116	        82.8 ns/op
BenchmarkMostSpecificHostMatch/50NeedleWildcard-8           	 3194504	       374 ns/op
BenchmarkMostSpecificHostMatch/100NeedleWildcard-8          	 1622061	       739 ns/op
BenchmarkMostSpecificHostMatch/1000NeedleWildcard-8         	  133128	      9025 ns/op
BenchmarkMostSpecificHostMatch/5000NeedleWildcard-8         	   48772	     24460 ns/op
PASS
ok  	command-line-arguments	22.478s
```

After use map:
```shell
➜ go test ./pilot/pkg/model/config_test.go -test.run MostSpecificHostMatch -bench=.
goos: darwin
goarch: amd64
BenchmarkMostSpecificHostMatch/10Exact-8         	88772878	        13.3 ns/op
BenchmarkMostSpecificHostMatch/50Exact-8         	67070103	        18.0 ns/op
BenchmarkMostSpecificHostMatch/100Exact-8        	65896960	        18.1 ns/op
BenchmarkMostSpecificHostMatch/1000Exact-8       	88457869	        13.6 ns/op
BenchmarkMostSpecificHostMatch/5000Exact-8       	61808491	        19.4 ns/op
BenchmarkMostSpecificHostMatch/10DestRuleWildcard-8         	18732081	        63.9 ns/op
BenchmarkMostSpecificHostMatch/50DestRuleWildcard-8         	 7212934	       176 ns/op
BenchmarkMostSpecificHostMatch/100DestRuleWildcard-8        	 2271836	       525 ns/op
BenchmarkMostSpecificHostMatch/1000DestRuleWildcard-8       	  234286	      5092 ns/op
BenchmarkMostSpecificHostMatch/5000DestRuleWildcard-8       	   80901	     14821 ns/op
BenchmarkMostSpecificHostMatch/10NeedleWildcard-8           	15178136	        78.5 ns/op
BenchmarkMostSpecificHostMatch/50NeedleWildcard-8           	 4088073	       291 ns/op
BenchmarkMostSpecificHostMatch/100NeedleWildcard-8          	 2141434	       560 ns/op
BenchmarkMostSpecificHostMatch/1000NeedleWildcard-8         	  221073	      5456 ns/op
BenchmarkMostSpecificHostMatch/5000NeedleWildcard-8         	  131659	      9112 ns/op
PASS
ok  	command-line-arguments	20.309s
```
You can see that using map is always faster than using slice whether small or big input sizes. In fact, loop a map is slower than slice. So except exact matching, we still use slice.

```golang
for _, h := range stack {
// ...
}
```

Reply:
@howardjohn Thanks for you reply first, and I think we can provide a parameter configuration for choosing an appropriate  inner data structures between a slice and a map.